File: 39-combination-sum/39-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    const ans = [];
    const seen = new Set;
    
    function backtrack(combination,target){
        if(target===0){
            combination.sort()
            if(!seen.has(combination.toString())){
                seen.add(combination.toString())
                ans.push(combination);
            }
            return;
        }
        
        if(target<0){
            return;
        }
        
        for(let candidate of candidates){
            backtrack([...combination,candidate], target - candidate);
        }
    }
    backtrack([],target);
    return ans;
};

File: 39-combination-sum/39-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res=[]
        
        def _backtrack(combo,idx):
            amt=sum(combo)
            if idx > len(candidates)-1: return
            if amt>target: return
            if amt==target:
                res.append(combo)
                return
            
            _backtrack([*combo,candidates[idx]],idx)
            _backtrack([*combo],idx+1)
            
        _backtrack([],0)
        return res
            